File: NeetCode 150/Heap/295. Find Median from Data Stream.py
class MedianFinder:

    def __init__(self):
        self.low = [] # max-heap (store -value)
        self.high = [] # min-heap
        

    def addNum(self, num: int) -> None:
        # push to low
        heapq.heappush(self.low, -num)

        # move the largest from low to high so that all low <= all high
        val = -heapq.heappop(self.low)
        heapq.heappush(self.high, val)

        # keep sizes balanced so low can have one extra element at most
        if len(self.high) > len(self.low):
            moved = heapq.heappop(self.high)
            heapq.heappush(self.low,-moved)
        
    def findMedian(self) -> float:
        if len(self.low) > len(self.high):
            return float(-self.low[0])
        else:
            return (-self.low[0] + self.high[0]) / 2.0
        


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: NeedCode 150 (JAVA)/Heaps/295. Find Median from Data Stream.java
class MedianFinder {
    private PriorityQueue<Integer> low; //max Heap
    private PriorityQueue<Integer> high; //min Heap

    public MedianFinder() {
        low = new PriorityQueue<>(Collections.reverseOrder());
        high = new PriorityQueue<>();
    }
    
    public void addNum(int num) {

        low.offer(num);

        high.offer(low.poll());

        if(low.size() < high.size()){
            low.offer(high.poll());
        }
        
    }
    
    public double findMedian() {

        if(low.size() > high.size())
            return low.peek();
        else
            return (low.peek() + high.peek()) / 2.0 ;    
    }
}

/*
Follow up:

If all integer numbers from the stream are in the range [0, 100], how would you optimize your solution?
If 99% of all integer numbers from the stream are in the range [0, 100], how would you optimize your solution?

For the first follow-up:

Since all values are in a small fixed range, I would replace the two heaps with a frequency array of size 101 and compute the median by scanning counts.

For the second:

I would use a hybrid solution: a counting array for the common [0,100] values and heaps or another ordered structure for the rare outliers.


*/

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
